### PR TITLE
[Role/Permission] Configuration tool (2 of 4)

### PR DESCRIPTION
**Pull request by ridz1208**
This adds the Role infrastructure to LORIS. This is one of 4 PRs which will ultimately result in assigning roles to users. each role has affiliated permissions and automatically checks the proper permission when selected. This change will be mainly visible in the user_accounts module.

PART 2: Add tool used to configure the role/permission relationships as well as populate the data  in the right table given an existing populated database.

in order to test, the following SQL inserts can be used:
```
INSERT INTO role (Name, Label) VALUES ('coordinator', 'Coordinator');
INSERT INTO role (Name, Label) VALUES ('data', 'Data-Entry');
INSERT INTO role (Name, Label) VALUES ('upload', 'Data-upload');


INSERT INTO role_permission_rel VALUES ((SELECT RoleID FROM role WHERE Name='coordinator'), (SELECT permID FROM permissions WHERE code='send_to_dcc'));
INSERT INTO role_permission_rel SELECT RoleID,permID FROM role JOIN permissions WHERE Name='data' AND code IN ('data_dict_view','data_entry','data_dict_edit','candidate_parameter_view');
INSERT INTO role_permission_rel SELECT RoleID,permID FROM role JOIN permissions WHERE Name='upload' AND code IN ('video_upload','media_write','media_read');
```

Original PR by @taracampbell : https://github.com/aces/Loris/pull/2642
Updated/Standardized/DAO full version : https://github.com/ridz1208/Loris/pull/9

### CODE REVIEW

**Comment by ridz1208** [comment]
Waiting on: https://github.com/aces/Loris/pull/3537

File: tools/role_permission_configurator.php [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+                echo "There are currently no roles in the database.\n";
+            }
+            exit(2);
+        case "permissions":
+            // User asking for the list of permissions
+
+            echo "The permissions in the database are:\n\n";
+            prettyPrint($permissions);
+            exit(3);
+        case "rebuildRoles":
```
Comment (PapillonMcGill) [inline tools/role_permission_configurator.php]: exit code other than 0 are normally for error or abnormal ending. When there is no error or problem, exit code is normally 0.

Could also use a single exit() call with a variable.

File: tools/role_permission_configurator.php [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+    // Insert the permission into permission role rel table
+    $DB->insert(
+        'role_permission_rel',
+        array(
+            'RoleID'       => $roleID,
+            'PermissionID' => $permissionID,
+        )
+    );
+
+    echo "\n$permission was added to the $role category.\n\n";
```
Comment (PapillonMcGill) [inline tools/role_permission_configurator.php]: "category" should be  "role" as other messages

File: tools/role_permission_configurator.php [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+ * Adds a permission to a role. Updates users with that role to either obtain
+ * the new permission or loose the role depending on the $update parameter
+ *
+ * @param string $role       the name of the role
+ * @param string $permission the name of the permission
+ * @param bool   $update     if true, adds permission to users with current role.
+ *                           if false, removes the role from the user
+ *
+ * @return void
+ */
```
Comment (PapillonMcGill) [inline tools/role_permission_configurator.php]: I don't see why a "addPermission" function would remove a role from a user.

This portion should be in a "removeRole" function

File: tools/role_permission_configurator.php [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+    $DB->delete(
+        'role_permission_rel',
+        array(
+            'RoleID'       => $roleID,
+            'PermissionID' => $permissionID,
+        )
+    );
+
+    echo "\n$permission was removed from the $role category.\n";
+
```
Comment (PapillonMcGill) [inline tools/role_permission_configurator.php]: again "category" instead of "role"

File: tools/role_permission_configurator.php [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+            'UID' => $userID,
+            'RID' => $roleID,
+        )
+    );
+
+    if (empty($result)) {
+        return false;
+    }
+    return true;
+}
```
Comment (PapillonMcGill) [inline tools/role_permission_configurator.php]: multiple return

**Comment by driusan** [comment]
Closing until #3537 goes in, until then it's going to need continual rebasing and just make it look like our PRs aren't being worked on.

File: tools/role_permission_configurator.php (hunk 3 above)
Comment (PapillonMcGill) [reply tools/role_permission_configurator.php]: Ok, now I understand, maybe I am slow, or maybe the text was not clear.

**Pull request closed without merging**